날씨 앱 [STEP 3-1] Blu, Jenna

calledBlu opened the pull request:
>안녕하세요! 코리(@corykim0829)🐶!
🌙**TEAM LuNna**🌙 의 Blu, Jenna입니다!
>
>STEP 3도 저번 스텝과 같이 1, 2로 분리해서 진행해 보려고 합니다.
>3-1에서는 Collection View를 활용하여 기초적인 앱의 UI 구성과 API로부터 받아온 데이터를 View에 적용하는 작업을 진행하였고,
>3-2에서는 요구사항 중 하나인 Configuration 부분에 대해 추가적으로 학습 후 활용하여 UI를 구현, 화면을 위에서 아래로 끌어당기는 경우 새로고침하는 기능을 구현할 예정입니다.
>
>아직 정리되지 못한 코드들이 많습니다.
>해당 부분은 코리의 리뷰를 적극 반영하여 STEP 3-2를 진행하면서 최대한 정리해보도록 하겠습니다!!
>
>먼저 연락 주신 덕분에 더 열심히 할 수 있었습니다!👍
>비가 오더니 날씨가 꽤 쌀쌀해진 거 같아요~! 감기 조심하세요 코리! 😷
>이번 리뷰도 잘 부탁드려요! 감사합니다! 🙇🏻🙇🏻


<br>

# 🧑🏼‍💻 작업 목록
## UI 구현(1)
- [x] Collection View를 활용하여 앱의 UI 구현
    - [x] header view에 현재 날씨 적용
    - [x] cell에 5일 예보 적용
- [X] 코드베이스 Auto Layout 구현
- [x] View에 데이터가 보이도록 적용
    - [x] DateFormatter를 활용하여 날짜 포맷 적용
- [x] 아이콘 이미지를 불러오도록 URLService 모델 변경 

<br>

![실행화면](https://user-images.githubusercontent.com/71758542/230283216-a06e55a3-3c93-4548-96fc-fe25efefbd6d.gif)

<br>


# 📚 학습 키워드 및 학습 내용
## 1. CollectionView
> 🖇️　참고한 공식문서: 
> - [Collection Views](https://developer.apple.com/documentation/uikit/views_and_controls/collection_views)
> - [UICollectionView](https://developer.apple.com/documentation/uikit/uicollectionview)
> - [UICollectionViewDataSource](https://developer.apple.com/documentation/uikit/uicollectionviewdatasource)
> - [UICollectionViewFlowLayout](https://developer.apple.com/documentation/uikit/uicollectionviewflowlayout)

- **학습 내용**
    - `UICollectionView는` `UITableView`처럼 'ordered collection of data items'를 레이아웃대로 띄울 수 있는 뷰이되, 'ordered'의 방향이 세로 리스트형태로 국한되지 않아 grid처럼도 구현이 가능한 타입
    - `UITableViewCellStyle`과 같은 건 없지만, 최근 list cell 및 layout이 추가되는 등 구현의 편의성과 유연함이 더해지고 있음
- **적용 과정**
    - `UICollectionViewListcell`, `UICollectionViewListlayout`을 쓰고자 했으나, 기본 콜렉션뷰를 구현해본 적이 없으므로 기초부터 학습하기로 합의, 커스텀 셀과 `UICollectionViewFlowLayout`을 이용해 구현
    - `defaultCellConfiguration()`을 쓰기에 앞서 콜렉션뷰 자체의 구현에 집중하고자 헤더뷰용, 셀용 커스텀셀을 정의하여 사용
- **문제와 해결**
    |문제와 해결|내용|
    |:---:|:---|
    |**문제**|- 비동기 데이터를 셀 서브뷰에 할당하는 작업이 어려웠음 |
    |**해결**|- Task 블록을 통해 필요한 async-await, completion handler의 작업이 모두 끝난 후에 콜렉션뷰의 reloadData()작업이 이루어지도록 통제 |
    
- **아쉬운 점**
    - 요구사항인 configuration을 이용하지 않음 -> ⭐️Step 3-2에서 리펙토링 예정

## 2. DateFormatter
> 🖇️　참고한 공식문서: 
> - [Date](https://developer.apple.com/documentation/foundation/date)
> - [DateFormatter](https://developer.apple.com/documentation/foundation/dateformatter)
> - [Data Formatting Guide](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/DataFormatting/Articles/dfDateFormatting10_4.html#//apple_ref/doc/uid/TP40002369-SW1)


- **학습 내용**
    - `DateFormatter`는 날짜와 텍스트 표현 사이의 변환기 역할을 함
    - 다양한 지역화된 사전 설명 및 구성 옵션을 제공
    - 날짜 및 시간을 고정 형식으로 표현하고 싶을 때 사용자 지정 형식 문자열을 지정할 수 있음
- **적용 과정**
    - String extension에 이니셜라이저로 구현
- **문제와 해결**
    |문제와 해결|내용|
    |:---:|:---|
    |**문제**|- `Date(timeIntervalSinceReferenceDate:)` 사용 시 월/일은 일치하지만 요일이 다르게 나타나는 문제 발생<br> - 한글로 된 요일을 얻기 위해 DateFormatter `setLocalizedDateFormatFromTemplate(_:)` 메서드를 활용하여 지정된 locale을 사용하여 날짜 형식을 설정하려고 하였지만 슬래시`/`가 제대로 표시되지 않는 문제 발생|
    |**해결**|- API의 dt에 해당하는 데이터가 unix, UTC로 unix UTC는 1970년 1월 1일 00:00:00 UTC 이후의 초 수로 시간을 표시하여 타임스탬프를 나타내는 방법이기 때문에 이에 해당하는 `Date(timeIntervalSince1970:)` 생성자를 사용하는 것으로 문제 해결 <br> - 공식문서상 `setLocalizedDateFormatFromTemplate(_:)` 메서드는 말 그대로 날짜 형식 패턴을 포함하는 문자열인 템플릿을 기반으로 데이터를 설정하는 것이기 때문에 "MM"이나 "h"으로 구성된 날짜 형식 패턴 템플릿을 넣어주면 자동으로 지정된 locale에 맞추어 localized해 주지만 이로 반환된 데이터는 프로젝트에서 요구하는 형태가 아님, 형태를 맞추기 위해 custom format을 지정해줄 수 있는 dateFormat 프로퍼티를 사용하여 "MM/dd(EE) HH시"로 지정해 줌

    
<br>

# 🙋🏻 질문 거리
## 1. 기본 타입(String, Double)의 Extension 관련
String extension(/Model/Extension/String+.swift)에 CoreLocation을 import, DateFormatter를 활용하여 아래 세가지 초기화자를 구현하였습니다.
1. Int 타입의 데이터를 String 타입으로 변환하는 초기화자
2. 위경도를 기반으로 변환된 주소 형태인 `CLPlacemark`를 인간 친화적인 주소로 변환하는 초기화자
3. Double 타입의 기온을 소수점 아래 한 자리만 남기고 도에 해당하는 기호(°)룰 붙여서 String 타입의 데이터로 변환하는 초기화자
구현하다보니 String의 extension 내부에 너무 많은 기능을 구현한 것이 아닌지 고민되었습니다.
지금처럼 진행해도 괜찮은지, 해당 역할을 하는 타입을 별도로 구현하는 것이 더 좋은지 질문드립니다!

## 2. URLService 네이밍 관련
`Weather`, `URL`, `Service`, `Request`, `Parse`, `Endpoint` 등의 키워드를 중복 없이 네이밍 가이드에 맞도록 적절히 배치하기 어려웠습니다. Service가 포괄적인 의미의 단어라고 들어 그리 적절치는 않은 이름이라고 생각했지만,
이미지 관련된 로직을 묶어 분리하기, 혹은 쌍으로 존재하는 parse·make·request류 네이밍의 메서드들을 통합하는 등의 리펙토링을 고려하고 있기에 && 따라서 앞으로 네이밍 근거가 변경될 여지가 있다고 판단하였기에 우선 현재 기능들을 고려하여 URL'Service'라고 네이밍 하였습니다.
혹시 추천하실 만한 네이밍 키워드가 있다면 조언 부탁드립니다...!

<br><br>

calledBlu committed:
[feat] Data Transfer Object 타입 구현

- ForecastWeatherData, CurrentWeatherData 구조체로 구현
- API로 송신한 데이터 중 View에 반영할 데이터로 구성된 DTO 구현

ueunli committed:
[feat] String extension 구현 및 DTO 수정

- String extension으로 이니셜라이저 3개 구현(일시, 기온, 장소를 문자열로 변환)함으로써 해당하는 정보가 String형태, 즉 DTO 내부 프로퍼티 타입으로 변환되어 담길 수 있도록 함
- ForecastWeatherData의 location프로퍼티를 삭제, CurrentWeatherData의 DataTime을 String?타입으로 설정하여 두 DTO를 WeatherData 하나로 통합
- ForecastWeatherComponents의 dataTime 프로퍼티에 String이 아닌 Int값이 담기도록 CodingKey 수정

calledBlu committed:
[refactor] 아이콘 API 호출 관련 메서드 추가 및 네이밍 변경

- WeatherURL -> URLService 네이밍 변경
- 네이밍 변경에 따른 request 관련 메서드를 WeatherParser로 이동
- 날씨 아이콘 API를 호출하기 위한 메서드 구현 및 기존 날씨 데이터 생성 메서드명 변경

ueunli committed:
[feat] 콜렉션뷰 및 필요한 부수타입 생성

- ViewController에 collectionView프로퍼티 생성
- ViewController가 collectionView의 delegate 및 dataSource로서 기능하도록 필수 메서드 구현
- HeaderView, CurrentWeatherCell, ForecastWeatherCell 정의

calledBlu committed:
[feat] collection view 생성 시 필요한 ForecastWeatherCell 내부 구현

- 주어진 UI와 동일하게 UILabel, UIImage를 추가하여 내부 구현
- 각 레이블이 임의의 값을 가지고 있도록 ViewController 수정
- HeaderView의 var로 되어있는 프로퍼티를 상수로 전환

ueunli committed:
[feat] DTO 이니셜라이저 구현 및 콜렉션뷰에 데이터 매칭

- 위치가 업데이트 되면 불러오던 날씨 데이터를 DTO로 변환하기 위한 생성자를 구현
- 뷰컨트롤러가 DTO형태의 날씨 데이터를 프로퍼티로 갖도록 수정
- current, forecast weather를 불러와 가공하는 로직을 각각 메서드로 묶어 분리
- DTO의 icon속성의 네이밍을 iconCode로 변경, UIImage?타입의 iconImage속성을 새로 정의
- DTO에 iconCode로부터 iconImage에 들어갈 값을 구하는 메서드, 최저&최고 기온을 한 문자열로 반환하는 메서드 구현
- URLService의 baseURL을 용도(날씨 데이터/아이콘)에 따라 2가지로 구분
- URLService가 어플리케이션 최적 크기의 아이콘을 가져올 수 있도록 이미지 명에 해당하는 path 수정
- HeaderView의 임시 이미지 제거

calledBlu committed:
[feat] 현재 좌표를 기반으로 주소로 변환하는 메서드 구현 및 view에 적용

- String extension으로 CLGeocoder로 좌표를 통해 가져온 주소 중 일부를 뷰에 알맞게 가공하는 initializer 구현
- view에 보여줄 사용자의 주소를 가지고 있는 속성인 userAddress를 View Controller에 선언
- initializer를 통해 변환된 주소가 view에 반영되도록 업데이트하는 메서드 구현 및 호출

ueunli committed:
[chore] 편의를 위해 부여했던 view별 배경색상 삭제

ueunli committed:
[chore] 불필요한 코드 제거

corykim0829 commented on WeatherForecast/WeatherForecast/Controller/ViewController.swift:
```diff
@@ -9,13 +9,79 @@ import CoreLocation
+        collectionView.delegate = self
+        view.addSubview(collectionView)
+        registerCollectionViewCell()
+        configureAutoLayout()
```
viewDidLoad와 같이 중요한 생명주기 코드가 복잡하면 가독성이 떨어지기 때문에
코드들을 함수로 추출해서 정리하면 좋을 것 같습니다.

calledBlu replied:
가독성을 좋게 하기 위해 기능별로 메서드를 분리하고 이들을 호출하는 메서드를 생성해 보았습니다!

> 해당 커밋: aec9f5e

corykim0829 commented on WeatherForecast/WeatherForecast/Controller/ViewController.swift:
```diff
@@ -9,13 +9,79 @@ import CoreLocation
     }
 }
 
+extension ViewController: UICollectionViewDelegateFlowLayout {
```
extension으로 구분한 부분 좋습니다 👍
지금은 괜찮지만, 추후에 ViewController 커지면 한 파일에 코드가 너무 많아서 가독성이 떨어지기 때문에, extension도 따른 파일로 분리할 수 있으면 분리하는게 좋습니다! 참고하세요 😊

corykim0829 commented on WeatherForecast/WeatherForecast/Controller/ViewController.swift:
```diff
@@ -9,13 +9,79 @@ import CoreLocation
 
+extension ViewController: UICollectionViewDelegateFlowLayout {
+    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, referenceSizeForHeaderInSection section: Int) -> CGSize {
+        return CGSizeMake(view.frame.width, 100)
```
사이즈 관련 매직넘버들을 정리하면 좋을 것 같아요 😊

calledBlu replied:
STEP 3-2에서 UICollectionViewCompositionalLayout의 list메서드를 사용 예정이라 해당 프로퍼티는 삭제될 예정입니다 😄
해당 부분 진행 시 사이즈 관련 매직넘버 정리에 유의하도록 하겠습니다! 리뷰 감사합니다 👍

corykim0829 commented on WeatherForecast/WeatherForecast/Controller/ViewController.swift:
```diff
@@ -9,13 +9,79 @@ import CoreLocation
+        cell.icon.image = forecastWeather?[indexPath.row].iconImage
+        cell.timeLabel.text = forecastWeather?[indexPath.row].dataTime
+        cell.temperatureLabel.text = forecastWeather?[indexPath.row].temperature
+        return cell
```
cell의 UI 요소들을 여기서 불러와서 값을 수정하는 것보다는
cell에 데이털르 넘겨서 cell 클래스 안에서 UI를 업데이트하는 구조로 변경하면 좋을 것 같아요.

ueunli replied:
셀 내 UI 요소들의 업데이트를 담당하는 메서드를 구현해서 아래와 같이 수정했어요!
```Swift
let cell = collectionView.dequeueReusableCell ...
cell.updateWeather(data)
return cell
```

> 해당커밋: 7c6050c

corykim0829 commented on WeatherForecast/WeatherForecast/Controller/ViewController.swift:
```diff
@@ -9,13 +9,79 @@ import CoreLocation
+        header.view.image.image = currentWeather?.iconImage
+        header.view.temperatureLabel.text = currentWeather?.temperature ?? "-"
+        header.view.minMaxTemperatureLabel.text = currentWeather?.temperatureString() ?? "-"
+        header.view.addressLabel.text = userAddress ?? "-"
```
여기도 위와 마찬가지로 데이터를 넘겨서 해당 클래스에서 UI를 업데이트하도록 변경해주세요 :)

ueunli replied:
위와 같이 셀 내에 메서드를 따로 구현하여 사용하는 방식으로 수정했어요!!

> 해당커밋: 7c6050c

corykim0829 commented on WeatherForecast/WeatherForecast/View/CurrentWeatherCell.swift:
```diff
@@ -0,0 +1,33 @@
+            view.leadingAnchor.constraint(equalTo: self.leadingAnchor),
+            view.trailingAnchor.constraint(equalTo: self.trailingAnchor),
+            view.heightAnchor.constraint(equalTo: self.heightAnchor)
+        ])
```
함수로 분리해서 처리해주세요!

ueunli replied:
`configureAutoLayout()` 메서드로 구현 및 호출했습니다 😃

> 해당커밋: 7c6050c

corykim0829 commented on WeatherForecast/WeatherForecast/View/CurrentWeatherCell.swift:
```diff
@@ -0,0 +1,33 @@
+    
+    required init?(coder: NSCoder) {
+        fatalError("Expected \(Self.self) initialization did fail")
+    }
```
fatalError로 처리하기보다는 super init 처리해서 위 init과 똑같이 처리해주세요 :)

ueunli replied:
아래와 같이 수정했어요! 

```Swift
required init?(coder: NSCoder) {
    super.init(coder: coder)
}
```

> 해당커밋: 7c6050c

corykim0829 replied:
기존 생성자와 동일하게 아래 부분 선언해주세요!
```swift
configureLayout()
temperatureLabel.textAlignment = .right
```

corykim0829 commented on WeatherForecast/WeatherForecast/View/CurrentWeatherCell.swift:
```diff
@@ -0,0 +1,33 @@
+    
+    override func prepareForReuse() {
+        super.prepareForReuse()
+        view.stackView.subviews.forEach { ($0 as? UILabel)?.text = nil }
```
forEach 는 개행을 해야 더 가독성이 좋을 것 같습니다 :)

corykim0829 replied:
여기도 함수로 빼서 정리하면 더 어떤 동작을 하는지 알기 좋을 것 같아요.

calledBlu replied:
> forEach 는 개행을 해야 더 가독성이 좋을 것 같습니다 :)

개행 수정 후 반영하였습니다!

> 해당 커밋: 6c17a23

corykim0829 commented on WeatherForecast/WeatherForecast/View/CurrentWeatherCell.swift:
```diff
@@ -0,0 +1,33 @@
+
+final class CurrentWeatherCell: UICollectionViewCell {
+    static let id = "current"
+    let view = HeaderView()
```
프로퍼티명을 단순히 view라고 하기보다는 headerView라고 명시하는게 좋을 것 같아요.

사실 Cell 안에 HeaderView가 명시되어있는게 어색하긴 해서, 여유가 있다면 HeaderView 자체를 등록해서 CollectionView에 사용하면 좋을 것 같아요!

calledBlu replied:
HeaderView를 CurrentWeatherView로, CurrentWeatherCell을 WeatherHeaderView로 어색한 부분을 개선해 보았고,
리뷰해주신 것처럼 프로퍼티명을 명시적으로 변경해 보았어요!

> 해당 커밋: 7c6050c

corykim0829 commented on WeatherForecast/WeatherForecast/Model/Extension/String+.swift:
```diff
@@ -0,0 +1,30 @@
+import CoreLocation
+
+extension String {
+    init(utcTime: Int) {
```
깔끔하게 잘 구현하셨네요 👍

corykim0829 commented on WeatherForecast/WeatherForecast/Model/Extension/String+.swift:
```diff
@@ -0,0 +1,30 @@
+    
+    init?(place: CLPlacemark) {
+        guard let administrativeArea = place.administrativeArea,
+              let thoroughfare = place.thoroughfare else { return nil }
```
return nil 개행 처리해주세요!

calledBlu replied:
개행 수정 후 반영하였습니다!

> 해당 커밋: 6c17a23

corykim0829 commented on WeatherForecast/WeatherForecast/View/ForecastWeatherCell.swift:
```diff
@@ -0,0 +1,62 @@
+    }
+    
+    required init?(coder: NSCoder) {
+        fatalError("Expected \(Self.self) initialization did fail")
```
여기 fatalError도 정리해주세요 :)

ueunli replied:
위와 마찬가지로 `super.init(coder: coder)`로 간단하게 수정했습니다 🙆🏻‍♀️

> 해당커밋: 7c6050c

corykim0829 commented on WeatherForecast/WeatherForecast/View/HeaderView.swift:
```diff
@@ -0,0 +1,69 @@
+
+import UIKit
+
+final class HeaderView: UIView {
```
단순히 HeaderView라고 하면 추상적이기 때문에 어떤 헤더뷰인지 클래스명을 바꾸면 좋을 것 같아요!

calledBlu replied:
앞서 Cell과 HeaderView의 네이밍을 변경하면서 `WeatherHeaderView`로 날씨에 대한 HeaderView임을 명시해 주었어요!
`CollectionViewHeaderView`로 네이밍 하는 것도 고민해 보았는데 View가 중복으로 들어가 가독성이 좋지 않을 거 같았어요 🥲

> 해당 커밋: 7c6050c

corykim0829 commented on WeatherForecast/WeatherForecast/Model/Network/WeatherParser.swift:
```diff
@@ -6,17 +6,40 @@
 import Foundation
+import UIKit
 
 struct WeatherParser<T: WeatherComposable> {
```
이제 실제 네트워크 요청도 하기 때문에 Parser라고 하기 보다는 더 어울리는 네이밍이 필요할 것 같아요!

ueunli replied:
조언 감사해요!! 
질문에 답변 주신 대로, 다음 스텝에서 해당 타입 리팩토링 후 내부 로직에 보다 알맞은 네이밍으로 수정해보도록 하겠습니다 🙆🏻‍♀️

corykim0829 reviewed (changes requested):
두분 정말 고생 많으셨습니다!!!

피드백 반영하고 다음 스텝 진행하면 좋을 것 같아요!

> 1. 기본 타입(String, Double)의 Extension 관련

개인적으로 타입 자체의 이니셜라이저로 처리하면 다양한 케이스를 통합 처리할 수 있어야한다고 생각이 듭니다. utcTime은 괜찮아보이는데 CLPPlacemark 같은 경우에는 꽤 구체적인 케이스라 CLPPlacemark의 extension 처리를 하는게 더 자연스러워 보이네요.

init(temperature: Double)는 굳이 extension까지는 필요가 없어 보이긴합니다!



> 2. URLService 네이밍 관련

URLService는 추상적인 것 같아요. 현재 해당 enum에서 결국 만들고 있는건 두가지 네트워크 요청을위한 URL을 만드는 작업을 하고 있기 때문에 모든 URL을 처리한다고 생각하기 보다는 해당 API의 URL을 생성한다는 느낌으로 네이밍을 하면 좋을 것 같습니다.

calledBlu committed:
[refactor] WeatherNetworkError 리팩토링

- 불필요한 invalidData case 삭제 및 invalidRequest -> requestFailed 네이밍 변경 및 연관값 설정
- description 속성을 통해 각 case별 error 부가설명을 제공하도록 구현

ueunli committed:
[refactor] api key 접근 관련 에러케이스 추가

- WeatherNetworkError에 `apiKeyNotFound`케이스 및 description 추가

calledBlu committed:
[chore] 가독성을 위한 개행 추가

- completion block, guard-else 구문의 로직 흐름을 명확히 하기 위한 개행 삽입

ueunli committed:
[refactor] 콜렉션뷰 셀 관련 리펙토링

- 네이밍 변경) HeaderView → CurrentWeatherView, CurrentWeatherCell → WeatherHeaderView
- 메서드 분리) 셀의 데이터 업데이트 로직을 셀 클래스 내부에서 수행하도록 변경 (updateData메서드)
- UICollectionViewCell 하위클래스의 init?(coder:) 내에서 fatalError메서드 대신 상위클래스 생성자를 호출하도록 변경
- prepareForReuse()메서드 내부의 셀 레이블 텍스트 초기화 기능을 별도 메서드로 분리

calledBlu committed:
[refactor] viewDidLoad 내부 코드를 메서드로 분리

- `configureLocationManager()` -> locationManager 관련 코드
- `configureCollectionView()` ->  CollectionView의 dataSource, delegate 할당 및 셀 등록 코드
- `configureAutoLayout()` -> CollectionView의 Auto Layout 관련 코드
